bodhi/server/consumers/pungi_wrapper.py

```diff
@@ -0,0 +1,348 @@
+            except Exception as ex:
+                self.compose.log_error("Couldn't create latest symlink: %s" % ex)
+
+
+class VariantsConfig(object):
+
+    """
+    This class generates a variants config which can be used with the compose
+    object to generate a dnf repo. Right now it only works with modules.
+    """
+
+    def __init__(self, updates, builds, variant_id="Server",
+                 arches=['x86_64', 'armhfp', 'aarch64', 'i386', 'ppc64', 'ppc64le', 's390x']):
+        """ TODO: to be defined1. """
+        self.updates = updates
+        self.builds = builds
+        self.modules = self._generate_module_list()
+        self.arches = arches
+        self.headers = [
+            '<?xml version="1.0" encoding="UTF-8"?>',
+            ('<!DOCTYPE variants PUBLIC "-//Red Hat, Inc.//DTD '
+             'Variants info//EN" "variants2012.dtd">'),
+        ]
+        self.body = [
+            '<variants>',
+            '<variant id="%s" name="%s" type="variant">' % (variant_id, variant_id),
+            '<arches>',
+            ''.join(['<arch>%s</arch>' % arch for arch in self.arches]),
+            '</arches>',
+            '<modules>',
+            ''.join(['<module>%s</module>' % module for module in self.modules]),
+            '</modules>',
+            '</variant>',
+            '</variants>'
+        ]
+
+    def _generate_module_list(self):
+        newest_builds = {}
+        # we loop through builds so we get rid of older builds and get only
+        # a dict with the newest builds
+        for build in self.builds:
+            nsv = build.nvr.rsplit('-', 1)
+            ns = nsv[0]
+            version = nsv[1]
+
+            if ns in newest_builds:
+                curr_version = newest_builds[ns]
+                if int(curr_version) < int(version):
+                    newest_builds[ns] = version
+            else:
+                newest_builds[ns] = version
+
+        # make sure that the modules we want to update get their correct versions
+        for update in self.updates:
+            nsv = update.title.rsplit('-', 1)
+            ns = nsv[0]
+            version = nsv[1]
+            newest_builds[ns] = version
+
+        module_list = ["%s-%s" % (nstream, v) for nstream, v in newest_builds.iteritems()]
+        return module_list
+
+    @property
+    def xml(self):
+        headers_str = "".join(self.headers)
+        body_str = "".join(self.body)
+        return str(headers_str + body_str)
+
+
+class PungiConfig(object):
+
+    """
+    Pungi config object which holds the main configuration file for pungi.
+
+    Attributes:
+        path (string) - this holds the path to the config file.
+    """
+
+    def __new__(cls, path, logger):
+        """ Reads the config from the provided path """
+        config = kobo.conf.PyConfigParser()
+        config.load_from_file(path)
+        cls._validate_conf(config, logger)
+        return config
+
+    @staticmethod
+    def _validate_conf(config, logger):
```

Comment:
Docblock. Also, why pass the logger instead of just making it a module variable?

Reply:
This appears to be gone, and thus fixed.